### PR TITLE
Add support for regex Patterns

### PR DESCRIPTION
**Pull request by aneeshusa**

### CODE REVIEW

**Comment by aneeshusa** [comment]
@ltworf I recently discovered typedload and find it a breath of fresh air compared to other libraries that don't work with e.g. standard dataclasses - thank you for making it!
I have been adopting it in my work codebases and noticed Patterns were not supported, hence the PR.
I looked at the CONTRIBUTING.md but didn't see any mention of updating the version or changelog - is that something I should do in the PR?

I see merging is blocked on getting the commit signed, I will look into how to set that up for this contribution.

**Comment by ltworf** [comment]
Hello, thanks for the contribution.

I'm on vacation now, I'll have a look at it next week I think.

File: typedload/typechecks.py [hunk 1]
```diff
@@ -254,6 +256,10 @@ def is_literal(type_: Any) -> bool:
     return getattr(type_, '__origin__', None) == Literal
 
 
+def is_pattern(type_: Any) -> bool:
```
Comment (ltworf) [inline typedload/typechecks.py:259]: This is fine, just lacking a docstring.

File: typedload/dataloader.py [hunk 2]
```diff
@@ -815,6 +817,14 @@ def _datetimeload(l: Loader, value: Any, type_) -> Union[datetime.date, datetime
         raise TypedloadValueError(str(e), type_=type_, value=value)
 
 
+def _patternload(l: Loader, value: Any, type_) -> re.Pattern:
+    if hasattr(type, "__args__"):
+        (input_type,) = type_.__args__
+        if input_type in {bytes, str} and type(value) != input_type:
+            raise TypedloadValueError('Got %s of type %s, expected %s' % (repr(value), tname(type(value)), tname(type_)), value=value, type_=type_)
+    return re.compile(value)
```
Comment (ltworf) [inline typedload/dataloader.py:825]: The re.compile might raise an exception. It seems it can be a TypeError or a re.error.

They have to be wrapped into typedload exceptions, to be able to later inspect where in the datatree the issue occurred.

File: typedload/dataloader.py [hunk 3]
```diff
@@ -222,8 +223,9 @@ def __init__(self, **kwargs) -> None:
             (is_literal, _literalload),
             (is_typeddict, _typeddictload),
             (lambda type_: type_ in {datetime.date, datetime.time, datetime.datetime}, _datetimeload),
-            (lambda type_: type_ == datetime.timedelta, _timedeltaload),
+            (is_pattern, _patternload),
             (lambda type_: type_ in self.strconstructed, _strconstructload),
+            (lambda type_: type_ == datetime.timedelta, _timedeltaload),
```
Comment (ltworf) [inline typedload/dataloader.py:228]: why did you move this line?

File: tests/test_dataloader.py [hunk 4]
```diff
@@ -462,6 +465,31 @@ def test_path(self):
         loader = dataloader.Loader()
         assert loader.load('/', Path) == Path('/')
 
+    def test_pattern_str(self):
+        loader = dataloader.Loader()
+        if sys.version_info[:2] <= (3, 8):
```
Comment (ltworf) [inline tests/test_dataloader.py:470]: This is not necessary, python 3.8 is the lowest supported version at this time.

File: tests/test_dataloader.py [hunk 5]
```diff
@@ -462,6 +465,31 @@ def test_path(self):
...
+    def test_pattern_bytes(self):
+        loader = dataloader.Loader()
+        if sys.version_info[:2] <= (3, 8):
+            with self.assertRaises(TypeError):
+                assert loader.load(br'[bc](at|ot)\d+', re.Pattern[bytes])
+        else:
+            assert loader.load(br'[bc](at|ot)\d+', re.Pattern[bytes]) == re.compile(br'[bc](at|ot)\d+')
+        assert loader.load(br'[bc](at|ot)\d+', typing.Pattern[bytes]) == re.compile(br'[bc](at|ot)\d+')
+
+    def test_pattern(self):
```
Comment (ltworf) [inline tests/test_dataloader.py:486]: Could you test loading broken stuff as well?

Like `None`, `33`, `"(((((((("`

**Review by ltworf — COMMENTED** [review]
Hello,

overall it is good. I commented on some things that need fixing before this can be merged.


For the future, I'd prefer separate commits for typecheck, dumper, loader, and tests… or maybe the tests can be added in the same commit of the relative function.

Yes, you should add an entry in the changelog, and update to https://github.com/ltworf/typedload/blob/master/docs/supported_types.md is needed as well.

Don't worry about the signature, I can override the setting if you've never used it before.

File: tests/test_dataloader.py (hunk 4 above)
Comment (aneeshusa) [reply tests/test_dataloader.py:470]: I could remove the `<` part if you like but when removing this entirely the tests fail on 3.8:
```
======================================================================
ERROR: test_pattern_bytes (tests.test_dataloader.TestCommonTypes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/tmp/src/typedload/tests/test_dataloader.py", line 485, in test_pattern_bytes
    assert loader.load(br'[bc](at|ot)\d+', re.Pattern[bytes]) == re.compile(br'[bc](at|ot)\d+')
TypeError: 'type' object is not subscriptable

======================================================================
ERROR: test_pattern_str (tests.test_dataloader.TestCommonTypes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/tmp/src/typedload/tests/test_dataloader.py", line 475, in test_pattern_str
    assert loader.load(r'[bc](at|ot)\d+', re.Pattern[str]) == re.compile(r'[bc](at|ot)\d+')
TypeError: 'type' object is not subscriptable
```

File: typedload/dataloader.py (hunk 3 above)
Comment (aneeshusa) [reply typedload/dataloader.py:227]: I think it was a remnant of earlier debugging but it should not be needed, undid this, thanks for pointing it out.

**Review by aneeshusa — COMMENTED** [review]
I hope you had a relaxing vacation! Thank you for the review.

Good point that this needs error handling, added that with tests. Updated the two docs you mentioned as well.

I addressed all the feedback in a new fixup-style commit; happy to squash it into the previous commit or squash into 3 commits (loader changes + tests, dumper changes + tests, docs changes), let me know what you prefer.

I would appreciate if you could override the setting about the signatures.

Let me know if there is anything I missed!

**Comment by ltworf** [comment]
Ok, merged.

Let me know if you'd like to contribute more, then I could add you to the project so you can do PR directly.

Thanks :)

**Comment by aneeshusa** [comment]
Thanks! I haven't run into any other meaningful pieces to contribute yet but will let you know if I do.

That being said, one minor note, I just noticed a few more places that may need to be updated to mention the new support from this PR:
- List of supported types in the README: https://github.com/ltworf/typedload/blob/master/README.md#supported-types
- List of supported types in package docstring: https://github.com/ltworf/typedload/blob/e77376962aeac6092b7613b4eda2010e95f914a0/typedload/__init__.py#L54-L75

I wonder if we need to have the list of supported types in 3 places and if we could consolidate some of those? In any case happy to make a PR tomorrow with whatever your preference is.

**Comment by ltworf** [comment]
Ah yes those should be updated as well, you want to make another pr for that?

Consolidation would be good, but they appear in different spots.

The readme is the "homepage" and also the text that goes on pypi, the docstring is converted on the website but also appears with `help()` in python, and the supported_types i guess is the more specific documentation with all the details. I don't have any good solution to merge the 3 without making the readme overly verbose.
